### ISSUE TITLE
Route tasks based on available host resources

### ISSUE THREAD

**Issue by Chris7** [celery/celery#4339]
Apologies in advance if this is covered elsewhere, but I couldn't find any reference that covered this use case.

I would like to specify task resource constraints in my task decorator (ie this task needs at least 0.3 cores, 2G memory, 50G disk, etc. etc.) and delay task acceptance by workers until there is sufficient resources. I currently manage this by using queues specific to hosts/workers after seeing what is the "best" host to send a task to, but the best host may actually change once the task is ready to execute. What I am looking for is when a task is read off the broker, a way to designate whether there is any worker suitable for the task, and if not, for the task to be placed back on the broker for a duration.

**Comment by georgepsarakis** [comment]
I think I understand what you mean, but I do not see any reliable way of achieving this though. The worker must check that itself meets all resource usage preconditions and then reject/accept the task. Aren't there many cases that some tasks are never being processed although they could be, because a host has approximately sufficient resources?

**Comment by thedrow** [comment]
Thank you for taking interest in Celery and bringing new feature ideas to us.
I think that this feature is out of scope for Celery.

I suggest that you use Celery to start a kuberenetes job using their API and specify your resource constrains there.
See https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/

**Comment by thedrow** [comment]
I'm going to close this issue for now but I'll be glad to revisit if you have a compelling argument to introduce such a feature to Celery.

**Comment by Chris7** [comment]
Thanks for getting back to me. This isn't a feature request (I completely understand it is way out of scope of something that celery would implement, but I do believe the primitives exist in celery to make this sort of functionality) -- this is more of a guidance request. 

From poking around I do see some relevant primitives exist that would allow a suboptimal version of this within celery (like `rusage` in app control, being able to call `self.retry()` if criteria is not met on the current worker). I also see discussion about the use of blueprints/etc. to customize and extend celery workers, but can't find any examples on how to properly control the task scheduler. So to summarize, I was hoping to learn whether there was a place that determined whether a worker should receive a task or if this is something that is fundamentally incompatible with celery (I get the sense it might just rely on rabbit delivering work and there is no notion of a "free worker" or the available pool of workers).

I really do like the k8s idea though and will check that out.
